homeworks/solutions/torantie/5-remesh/src/halfedge.js

```diff
@@ -0,0 +1,639 @@
+    this.edgeQuadric = vertexQuadric.add(nextVertexQuadric)
+
+  }
+
+
+  getQuadricError(edgeQuadric, targetVertexPosition){
+    //(Q1+Q2) * x
+    let v = edgeQuadric.mulpos(targetVertexPosition)
+    let h = edgeQuadric.x30 * targetVertexPosition.x
+        + edgeQuadric.x31 * targetVertexPosition.y
+        + edgeQuadric.x32 * targetVertexPosition.z
+        + edgeQuadric.x33
+
+    let edgeError = v.x * targetVertexPosition.x
+        + v.y * targetVertexPosition.y
+        + v.z * targetVertexPosition.z
+        + h
+
+    return edgeError
+  }
+
+  collapse(){
+    let vertexToDelete = this.halfedge.twin.vertex
+
+    let halfedgesToDelete = []
+
+    halfedgesToDelete.push(this.halfedge)
+    halfedgesToDelete.push(this.halfedge.next)
+    halfedgesToDelete.push(this.halfedge.next.next)
+    halfedgesToDelete.push(this.halfedge.twin)
+    halfedgesToDelete.push(this.halfedge.twin.next)
+    halfedgesToDelete.push(this.halfedge.twin.next.next)
+
+    let edgesToDelete = []
+
+    edgesToDelete.push(this)
+    edgesToDelete.push(this.halfedge.next.edge)
+    edgesToDelete.push(this.halfedge.twin.prev.edge)
+
+    edgesToDelete.forEach((e) =>{
+      e.isDeleted = true
+    })
+
+    let facesToDelete = []
+    facesToDelete.push(this.halfedge.face)
+    facesToDelete.push(this.halfedge.twin.face)
+    facesToDelete.forEach((f) =>{
+      f.isDeleted = true
+    })
+
+    //change vertex position
+    let vertex = this.halfedge.vertex
+    vertex.position = this.targetVertexPosition
+    vertex.halfedge = this.halfedge.next.twin.next
+
+    // set the vertex with the new vertex position as vertex of all halfedges attached to the vertex that will be deleted
+    vertexToDelete.halfedges((halfedge)=>{
+      halfedge.vertex = vertex
+    })
+
+    vertexToDelete.isDeleted = true
+
+    //set edge references: right halfedges edge reference should be the left edges, since we delete the edges on the right
+    this.halfedge.next.twin.edge = this.halfedge.prev.edge
+    this.halfedge.twin.prev.twin.edge = this.halfedge.twin.next.edge
+
+    //change vertex references
+    this.halfedge.prev.vertex.halfedge = this.halfedge.next.twin
+    this.halfedge.twin.prev.vertex.halfedge = this.halfedge.twin.next.twin
+
+    //edge merge: merge the 4 edges by setting their twins
+    this.halfedge.next.twin.twin = this.halfedge.prev.twin
+    this.halfedge.prev.twin.twin = this.halfedge.next.twin
+    //left up edge needs new halfedge as reference
+    if(this.halfedge.prev.idx == this.halfedge.prev.edge.halfedge.idx)
+      this.halfedge.prev.edge.halfedge = this.halfedge.next.twin
+
+    this.halfedge.twin.next.twin.twin = this.halfedge.twin.prev.twin
+    //left down edge needs new halfedge as reference
+    if(this.halfedge.twin.next.idx == this.halfedge.twin.next.edge.halfedge.idx)
+      this.halfedge.twin.next.edge.halfedge = this.halfedge.twin.prev.twin
+
+    this.halfedge.twin.prev.twin.twin = this.halfedge.twin.next.twin
```

Comment:
I think your approach is the ideal approach to do edge collapse: assign the target position to an existing vertex, then merge the other to the reassigned vertex.

But the approach requires a more careful handle:

1. two remaining face constructs a quad, and the edge to collapse is the diagonal edge. In this case, the result will become two edges (no face).
2. three triangles and the decimated edge is on the boundary: result in a single triangle.
3. detect if edge (v1-v2) collapse from v2 to v1 is still valid (detect flip faces). If not, skip this edge collapse operation and restore the position of v1.

I don't have a clear mind about if your implementation works on these cases. Do you still have your code in mind?

Reply:
1. I think case one should still be accounted for in my code since i never actually reassign faces but only edge references. 
2. Shouldn't the collapse be skipped if the edge to collapse is on the boundary?
3. I think that is accounted for by checking if the edge was already deleted and then skipping the edge collapse accordingly.
Sorry for the late reply. Had to re-read my code to get it back in mind^^